src/server/pps/server/api_server.go

```diff
@@ -945,68 +939,75 @@ func (a *apiServer) DeleteJob(ctx context.Context, request *pps.DeleteJobRequest
 func (a *apiServer) StopJob(ctx context.Context, request *pps.StopJobRequest) (response *types.Empty, retErr error) {
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
-	pachClient := a.env.GetPachClient(ctx)
-	if err := a.stopJob(ctx, pachClient, request.Job, request.OutputCommit, "job stopped"); err != nil {
+	if err := a.txnEnv.WithTransaction(ctx, func(txn txnenv.Transaction) error {
+		return txn.StopJob(request)
+	}); err != nil {
 		return nil, err
 	}
 	return &types.Empty{}, nil
 }
 
-func (a *apiServer) stopJob(ctx context.Context, pachClient *client.APIClient, job *pps.Job, outputCommit *pfs.Commit, reason string) error {
-	if job == nil && outputCommit == nil {
-		return errors.New("Job or OutputCommit must be specified")
-	}
+// StopJobInTransaction is identical to StopJob except that it can run inside an
+// existing postgres transaction.  This is not an RPC.
+func (a *apiServer) StopJobInTransaction(txnCtx *txnenv.TransactionContext, request *pps.StopJobRequest) error {
+	return a.stopJob(txnCtx, request.Job, request.OutputCommit, "job stopped")
+}
+
+func (a *apiServer) stopJob(txnCtx *txnenv.TransactionContext, job *pps.Job, outputCommit *pfs.Commit, reason string) error {
+	jobs := a.jobs.ReadWrite(txnCtx.SqlTx)
+	pipelineJobInfo := &pps.StoredPipelineJobInfo{}
 	if job != nil {
-		pipelineJobInfo, err := a.InspectJob(ctx, &pps.InspectJobRequest{Job: job})
-		if err != nil {
+		if err := jobs.Get(job.ID, pipelineJobInfo); err != nil {
 			return err
 		}
-		outputCommit = pipelineJobInfo.OutputCommit
-	}
-	commitInfo, err := pachClient.InspectCommit(outputCommit.Repo.Name, outputCommit.ID)
-	if err != nil {
-		if pfsServer.IsCommitNotFoundErr(err) || pfsServer.IsCommitDeletedErr(err) {
+	} else if outputCommit != nil {
+		count := 0
+		if err := jobs.GetByIndex(ppsdb.JobsOutputIndex, pfsdb.CommitKey(outputCommit), pipelineJobInfo, col.DefaultOptions(), func(string) error {
+			count++
+			if count > 1 {
+				return errors.Errorf("found multiple jobs with output commit: %s@%s", outputCommit.Repo.Name, outputCommit.ID)
+			}
 			return nil
+		}); err != nil {
+			return err
+		}
+		if count == 0 {
+			return errors.Errorf("job with output commit %s@%s not found", outputCommit.Repo.Name, outputCommit.ID)
```

Comment:
I'm actually moving towards using `pfsdb.CommitKey` for this (because it's less ambiguous and uses the same format we use on the command-line for user-friendliness), but that'll need to wait for global IDs to be fully done.